pkg/controller/nodelifecycle/node_lifecycle_controller.go

```diff
@@ -824,71 +830,103 @@ func (nc *Controller) monitorNodeHealth(ctx context.Context) error {
 		delete(nc.knownNodeSet, deleted[i].Name)
 	}
 
+	// Marking the pods not ready on a node requires looping over them and
+	// updating each pod's status one at a time. This is performed serially, and
+	// can take a while if we're processing each node serially as well. So we
+	// process them with bounded concurrency instead, since most of the time is
+	// spent waiting on io.
+	sem := semaphore.NewWeighted(scheduler.UpdateWorkerSize)
```

Comment:
A unit test is needed to check different worker sizes.

Reply:
Why would you pick semaphore over using a separate goroutine which could update pods statuses?

Reply:
The semaphore package makes it is easy to implement worker pools to work on independent tasks with bounded parallelism, but I'm not tied to it. Could you elaborate on what you had in mind with the separate goroutine approach?

Reply:
+1 to separate goroutines.
In general, the "multiple workers per controller" is a common pattern in many controllers. I think that making different controllers as similar as possible has a huge value for developers, as they don't need to learn a separate patterns for each thing they are touching.

Here the pattern is slightly different (i.e. you don't create N workers at the top of level of the controller), instead you create M workers for processing here.
But even for this pattern, we have a utility function. This function is `ParallelizeUntil`:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/util/workqueue/parallelizer.go#L46

I would really prefer changing the approach to use this function, instead of introducing semaphore here (which isn't yet used anywhere else in the code).


[At some point we may consider changing the implementation of ParallelizeUntil to use semaphore, but let's not introduce it here.]

Reply:
Updated to use `ParallelizeUntil`. It dropped in nicely, and also made the diff smaller.